No suggestions offered when matching on a String

### Code

```Rust
fn main() {
    let s = "yes".to_owned();
    
    let bool_equiv = match s {
        "yes" | "yeah" => Some(true),
        "no" | "nope" => Some(false),
        _ => None,
    };
    
    dbg!(bool_equiv);
}
```

### Current output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:5:9
  |
4 |     let bool_equiv = match s {
  |                            - this expression has type `String`
5 |         "yes" | "yeah" => Some(true),
  |         ^^^^^ expected `String`, found `&str`
```

### Desired output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:5:9
  |
4 |     let bool_equiv = match s {
  |                            - this expression has type `String`
5 |         "yes" | "yeah" => Some(true),
  |         ^^^^^ expected `String`, found `&str`
  |               help: Use .as_ref() to match on a String value
```

### Rationale and extra context

Both .as_ref() and .as_str() would be reasonable suggestions here I think.

### Rust Version

```Shell
$ rustc --version --verbose
rustc 1.97.0-nightly (ad3a598ca 2026-05-03)
binary: rustc
commit-hash: ad3a598ca4bc7c68bcbbce3e0d3be9a7618df190
commit-date: 2026-05-03
host: x86_64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.4
```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_typeck/src/pat.rs
+++ b/compiler/rustc_hir_typeck/src/pat.rs
@@ -1007,7 +1007,22 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         //
         // then that's equivalent to there existing a LUB.
         let cause = self.pattern_cause(ti, span);
-        if let Err(err) = self.demand_suptype_with_origin(&cause, expected, pat_ty) {
+        if let Err(mut err) = self.demand_suptype_with_origin(&cause, expected, pat_ty) {
+            // If scrutinee is String and pattern is &str, suggest .as_str()
+            let expected = self.resolve_vars_with_obligations(expected);
+            if let ty::Adt(adt, _) = expected.kind()
+                && self.tcx.is_lang_item(adt.did(), LangItem::String)
+                && pat_ty.is_ref()
+                && pat_ty.peel_refs().is_str()
+                && let Some(origin_expr) = ti.origin_expr
+            {
+                err.span_suggestion_verbose(
+                    origin_expr.span.shrink_to_hi(),
+                    "consider converting the `String` to a `&str` using `.as_str()`",
+                    ".as_str()",
+                    Applicability::MachineApplicable,
+                );
+            }
             err.emit();
         }
 

--- a/tests/ui/pattern/deref-patterns/needs-gate.stderr
+++ b/tests/ui/pattern/deref-patterns/needs-gate.stderr
@@ -76,6 +76,11 @@ LL |     match "str".to_owned() {
    |           ---------------- this expression has type `String`
 LL |         "str" => {}
    |         ^^^^^ expected `String`, found `&str`
+   |
+help: consider converting the `String` to a `&str` using `.as_str()`
+   |
+LL |     match "str".to_owned().as_str() {
+   |                           +++++++++
 
 error[E0308]: mismatched types
   --> $DIR/needs-gate.rs:52:12

--- a/tests/ui/pattern/string-match-as-str-suggestion.fixed
+++ b/tests/ui/pattern/string-match-as-str-suggestion.fixed
@@ -1,0 +1,18 @@
+//@ run-rustfix
+
+fn main() {
+    let s = "yes".to_owned();
+
+    let _ = match s.as_str() {
+        "yes" => Some(true),
+        //~^ ERROR mismatched types
+        "no" => Some(false),
+        //~^ ERROR mismatched types
+        _ => None,
+    };
+
+    let s2 = String::from("hello");
+    if let "hello" = s2.as_str() {
+        //~^ ERROR mismatched types
+    }
+}

--- a/tests/ui/pattern/string-match-as-str-suggestion.rs
+++ b/tests/ui/pattern/string-match-as-str-suggestion.rs
@@ -1,0 +1,18 @@
+//@ run-rustfix
+
+fn main() {
+    let s = "yes".to_owned();
+
+    let _ = match s {
+        "yes" => Some(true),
+        //~^ ERROR mismatched types
+        "no" => Some(false),
+        //~^ ERROR mismatched types
+        _ => None,
+    };
+
+    let s2 = String::from("hello");
+    if let "hello" = s2 {
+        //~^ ERROR mismatched types
+    }
+}

--- a/tests/ui/pattern/string-match-as-str-suggestion.stderr
+++ b/tests/ui/pattern/string-match-as-str-suggestion.stderr
@@ -1,0 +1,43 @@
+error[E0308]: mismatched types
+  --> $DIR/string-match-as-str-suggestion.rs:7:9
+   |
+LL |     let _ = match s {
+   |                   - this expression has type `String`
+LL |         "yes" => Some(true),
+   |         ^^^^^ expected `String`, found `&str`
+   |
+help: consider converting the `String` to a `&str` using `.as_str()`
+   |
+LL |     let _ = match s.as_str() {
+   |                    +++++++++
+
+error[E0308]: mismatched types
+  --> $DIR/string-match-as-str-suggestion.rs:9:9
+   |
+LL |     let _ = match s {
+   |                   - this expression has type `String`
+...
+LL |         "no" => Some(false),
+   |         ^^^^ expected `String`, found `&str`
+   |
+help: consider converting the `String` to a `&str` using `.as_str()`
+   |
+LL |     let _ = match s.as_str() {
+   |                    +++++++++
+
+error[E0308]: mismatched types
+  --> $DIR/string-match-as-str-suggestion.rs:15:12
+   |
+LL |     if let "hello" = s2 {
+   |            ^^^^^^^   -- this expression has type `String`
+   |            |
+   |            expected `String`, found `&str`
+   |
+help: consider converting the `String` to a `&str` using `.as_str()`
+   |
+LL |     if let "hello" = s2.as_str() {
+   |                        +++++++++
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

